File: api/groq.py
# api/groq.py - Groq API Proxy for Vercel Serverless Function

import os
import json
import urllib.request
import urllib.error

def handler(request):
    """
    Vercel Python Serverless Function handler.
    Receives POST requests from frontend, forwards to Groq API.
    """
    # CORS headers
    headers = {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Methods': 'POST, OPTIONS',
        'Access-Control-Allow-Headers': 'Content-Type',
    }

    # Handle preflight OPTIONS request
    if request.method == 'OPTIONS':
        return {
            'statusCode': 204,
            'headers': headers
        }

    # Only POST allowed
    if request.method != 'POST':
        return {
            'statusCode': 405,
            'headers': headers,
            'body': json.dumps({'error': 'Method not allowed. Use POST.'})
        }

    try:
        # Parse request body
        body = json.loads(request.body)
        messages = body.get('messages')
        model = body.get('model', 'openai/gpt-oss-120b')
        temperature = body.get('temperature', 0.7)
        max_tokens = body.get('max_tokens', 350)

        if not messages:
            return {
                'statusCode': 400,
                'headers': headers,
                'body': json.dumps({'error': 'Messages field is required'})
            }

        # ✅ Read API key from Vercel Environment Variable (SECURE)
        GROQ_API_KEY = os.environ.get('GROQ_API_KEY')
        if not GROQ_API_KEY:
            return {
                'statusCode': 500,
                'headers': headers,
                'body': json.dumps({'error': 'GROQ_API_KEY not configured on server.'})
            }

        # Prepare Groq API request
        groq_url = 'https://api.groq.com/openai/v1/chat/completions'
        groq_headers = {
            'Authorization': f'Bearer {GROQ_API_KEY}',
            'Content-Type': 'application/json'
        }
        groq_body = json.dumps({
            'model': model,
            'messages': messages,
            'temperature': temperature,
            'max_tokens': max_tokens,
            'top_p': 0.9
        }).encode('utf-8')

        # Send request to Groq
        req = urllib.request.Request(groq_url, data=groq_body, headers=groq_headers, method='POST')
        
        with urllib.request.urlopen(req) as response:
            response_data = json.loads(response.read().decode('utf-8'))
            reply = response_data.get('choices', [{}])[0].get('message', {}).get('content', '').strip()
            
            return {
                'statusCode': 200,
                'headers': headers,
                'body': json.dumps({'reply': reply})
            }

    except urllib.error.HTTPError as e:
        error_body = e.read().decode('utf-8')
        print(f'Groq API HTTP Error: {e.code} - {error_body}')
        return {
            'statusCode': e.code,
            'headers': headers,
            'body': json.dumps({'error': f'Groq API error: {error_body}'})
        }
    except Exception as e:
        print(f'Server Error: {str(e)}')
        return {
            'statusCode': 500,
            'headers': headers,
            'body': json.dumps({'error': str(e)})
        }

File: api/groq.js
// api/groq.js - Vercel Serverless Function (Node.js)

module.exports = async (req, res) => {
    // CORS headers
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

    // Handle preflight
    if (req.method === 'OPTIONS') {
        return res.status(204).end();
    }

    if (req.method !== 'POST') {
        return res.status(405).json({ error: 'Method not allowed' });
    }

    try {
        const { messages, model = 'openai/gpt-oss-120b', temperature = 0.7, max_tokens = 350 } = req.body;

        // ✅ API key Environment Variable থেকে পড়া হচ্ছে
        const GROQ_API_KEY = process.env.GROQ_API_KEY;
        if (!GROQ_API_KEY) {
            return res.status(500).json({ error: 'GROQ_API_KEY not configured' });
        }

        const response = await fetch('https://api.groq.com/openai/v1/chat/completions', {
            method: 'POST',
            headers: {
                'Authorization': `Bearer ${GROQ_API_KEY}`,
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                model,
                messages,
                temperature,
                max_tokens,
                top_p: 0.9
            })
        });

        const data = await response.json();
        if (!response.ok) {
            console.error('Groq API Error:', data);
            return res.status(response.status).json({ error: data });
        }

        const reply = data.choices?.[0]?.message?.content?.trim() || '';
        return res.status(200).json({ reply });

    } catch (error) {
        console.error('Proxy Error:', error);
        return res.status(500).json({ error: error.message });
    }
};
